plugins/fabric8-multi-tenant-manager/src/main/java/com/redhat/che/multitenant/tenantdata/TenantDataCacheLoader.java

```diff
@@ -0,0 +1,111 @@
+
+import static java.lang.String.format;
+
+import com.google.common.cache.CacheLoader;
+import com.google.gson.Gson;
+import com.google.gson.JsonSyntaxException;
+import com.google.inject.Inject;
+import com.google.inject.name.Named;
+import com.redhat.che.multitenant.tenantdata.UserServicesJsonResponse.Namespace;
+import java.io.IOException;
+import java.util.List;
+import org.eclipse.che.api.core.BadRequestException;
+import org.eclipse.che.api.core.ConflictException;
+import org.eclipse.che.api.core.ForbiddenException;
+import org.eclipse.che.api.core.NotFoundException;
+import org.eclipse.che.api.core.ServerException;
+import org.eclipse.che.api.core.UnauthorizedException;
+import org.eclipse.che.api.core.ValidationException;
+import org.eclipse.che.api.core.rest.HttpJsonRequestFactory;
+import org.eclipse.che.api.core.rest.HttpJsonResponse;
+import org.eclipse.che.api.workspace.server.spi.InfrastructureException;
+
+class TenantDataCacheLoader extends CacheLoader<TenantDataCacheKey, UserCheTenantData> {
+
+  private final boolean standalone;
+  private final String cheNamespace;
+  private final HttpJsonRequestFactory httpJsonRequestFactory;
+  private final String fabric8UserServiceEndpoint;
+
+  @Inject
+  TenantDataCacheLoader(
+      HttpJsonRequestFactory httpJsonRequestFactory,
+      @Named("che.fabric8.user_service.endpoint") String fabric8UserServiceEndpoint,
+      @Named("che.infra.openshift.project") String cheNamespace,
+      @Named("che.fabric8.standalone") boolean standalone) {
+    this.fabric8UserServiceEndpoint = fabric8UserServiceEndpoint;
+    this.cheNamespace = cheNamespace;
+    this.standalone = standalone;
+    this.httpJsonRequestFactory = httpJsonRequestFactory;
+  }
+
+  @Override
+  public UserCheTenantData load(TenantDataCacheKey cacheKey) throws InfrastructureException {
+    if (standalone) {
+      return loadStandalone(cacheKey);
+    }
+
+    String responseBody;
+    try {
+      responseBody = getResponseBody(fabric8UserServiceEndpoint, cacheKey.getKeycloakToken());
+    } catch (Exception e) {
+      throw new InfrastructureException("Exception during the user tenant data retrieval", e);
+    }
+    try {
+      final Gson gson = new Gson();
+      final UserServicesJsonResponse userServicesData =
+          gson.fromJson(responseBody, UserServicesJsonResponse.class);
+      List<Namespace> namespaces = userServicesData.getNamespaces();
+      for (Namespace ns : namespaces) {
+        if (cacheKey.getNamespaceType().equals(ns.getType())) {
+          UserCheTenantData cheTenantData =
+              new UserCheTenantData(
+                  ns.getName(),
+                  ns.getClusterUrl(),
+                  ns.getClusterAppDomain(),
+                  ns.isClusterCapacityExhausted());
+          UserCheTenantDataValidator.validate(cheTenantData);
+          return cheTenantData;
+        }
+      }
+    } catch (ValidationException | JsonSyntaxException e) {
```

Comment:
Is NPE possible here?

Reply:
Good point -- I was looking at the validator but if the response cannot be parsed at all then `ns` could be null.

Added test case to cover this scenario.